File: DiKErnel/Test/DiKErnel.Integration.Test/Internal/Data/ProfileDataTest.cpp
// Copyright (C) Stichting Deltares 2022. All rights reserved.
//
// This file is part of DiKErnel.
//
// DiKErnel is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, version 3.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
//
// All names, logos, and references to "Deltares" are registered trademarks of
// Stichting Deltares and remain full property of Stichting Deltares at all times.
// All rights reserved.

#include <gtest/gtest.h>

#include "ProfileData.h"

namespace DiKErnel::Integration::Test
{
    using namespace Core;
    using namespace std;
    using namespace testing;

    struct ProfileDataTest : Test
    {
        static unique_ptr<ProfileData> CreateDefaultProfileData()
        {
            auto pointOne = make_shared<ProfilePoint>(1.0, 1.1);
            auto pointTwo = make_shared<ProfilePoint>(2.0, 2.2);
            auto pointThree = make_shared<ProfilePoint>(3.0, 3.3);

            vector<unique_ptr<ProfileSegment>> testDikeProfileSegments;
            testDikeProfileSegments.push_back(make_unique<ProfileSegment>(pointOne, pointTwo, 1.0));
            testDikeProfileSegments.push_back(make_unique<ProfileSegment>(pointTwo, pointThree, 1.0));

            return make_unique<ProfileData>(move(testDikeProfileSegments), vector<unique_ptr<CharacteristicPoint>>());
        }
    };

    TEST_F(ProfileDataTest, InterpolationVerticalHeight_HorizontalPositionLeftOfDikeProfile_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto verticalHeight = profileData->InterpolationVerticalHeight(0.0);

        // Assert
        ASSERT_DOUBLE_EQ(numeric_limits<double>::infinity(), verticalHeight);
    }

    TEST_F(ProfileDataTest, InterpolationVerticalHeight_HorizontalPositionOnFirstDikeProfilePoint_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto verticalHeight = profileData->InterpolationVerticalHeight(1.0);

        // Assert
        ASSERT_DOUBLE_EQ(1.1, verticalHeight);
    }

    TEST_F(ProfileDataTest, InterpolationVerticalHeight_HorizontalPositionOnRandomDikeProfilePoint_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto verticalHeight = profileData->InterpolationVerticalHeight(2.0);

        // Assert
        ASSERT_DOUBLE_EQ(2.2, verticalHeight);
    }

    TEST_F(ProfileDataTest, InterpolationVerticalHeight_HorizontalPositionOnLastDikeProfilePoint_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto verticalHeight = profileData->InterpolationVerticalHeight(3.0);

        // Assert
        ASSERT_DOUBLE_EQ(3.3, verticalHeight);
    }

    TEST_F(ProfileDataTest, InterpolationVerticalHeight_HorizontalPositionBetweenAscendingDikeProfilePoints_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto verticalHeight = profileData->InterpolationVerticalHeight(1.5);

        // Assert
        ASSERT_DOUBLE_EQ(1.65, verticalHeight);
    }

    TEST_F(ProfileDataTest, InterpolationVerticalHeight_HorizontalPositionBetweenDescendingDikeProfilePoints_ExpectedValue)
    {
        // Setup
        auto pointOne = make_shared<ProfilePoint>(1.0, 3.3);
        auto pointTwo = make_shared<ProfilePoint>(2.0, 2.2);
        auto pointThree = make_shared<ProfilePoint>(3.0, 1.1);

        vector<unique_ptr<ProfileSegment>> testDikeProfileSegmentsWithDescendingHeight;
        testDikeProfileSegmentsWithDescendingHeight.push_back(make_unique<ProfileSegment>(pointOne, pointTwo, 1.0));
        testDikeProfileSegmentsWithDescendingHeight.push_back(make_unique<ProfileSegment>(pointTwo, pointThree, 1.0));

        const ProfileData profileData(move(testDikeProfileSegmentsWithDescendingHeight), vector<unique_ptr<CharacteristicPoint>>());

        // Call
        const auto verticalHeight = profileData.InterpolationVerticalHeight(2.5);

        // Assert
        ASSERT_DOUBLE_EQ(1.65, verticalHeight);
    }

    TEST_F(ProfileDataTest, InterpolationVerticalHeight_HorizontalPositionRightOfDikeProfile_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto verticalHeight = profileData->InterpolationVerticalHeight(4.0);

        // Assert
        ASSERT_DOUBLE_EQ(numeric_limits<double>::infinity(), verticalHeight);
    }

    TEST_F(ProfileDataTest, InterpolationHorizontalPosition_VerticalHeightBelowDikeProfile_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto horizontalPosition = profileData->InterpolationHorizontalPosition(0.0);

        // Assert
        ASSERT_DOUBLE_EQ(numeric_limits<double>::infinity(), horizontalPosition);
    }

    TEST_F(ProfileDataTest, InterpolationHorizontalPosition_VerticalHeightOnFirstDikeProfilePoint_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto horizontalPosition = profileData->InterpolationHorizontalPosition(1.1);

        // Assert
        ASSERT_DOUBLE_EQ(1.0, horizontalPosition);
    }

    TEST_F(ProfileDataTest, InterpolationHorizontalPosition_VerticalHeightOnRandomDikeProfilePoint_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto horizontalPosition = profileData->InterpolationHorizontalPosition(2.2);

        // Assert
        ASSERT_DOUBLE_EQ(2.0, horizontalPosition);
    }

    TEST_F(ProfileDataTest, InterpolationHorizontalPosition_VerticalHeightOnLastDikeProfilePoint_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto horizontalPosition = profileData->InterpolationHorizontalPosition(3.3);

        // Assert
        ASSERT_DOUBLE_EQ(3.0, horizontalPosition);
    }

    TEST_F(ProfileDataTest, InterpolationHorizontalPosition_VerticalHeightBetweenProfilePoints_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto horizontalPosition = profileData->InterpolationHorizontalPosition(1.65);

        // Assert
        ASSERT_DOUBLE_EQ(1.5, horizontalPosition);
    }

    TEST_F(ProfileDataTest, InterpolationHorizontalPosition_VerticalHeightAboveDikeProfile_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto horizontalPosition = profileData->InterpolationHorizontalPosition(4.0);

        // Assert
        ASSERT_DOUBLE_EQ(numeric_limits<double>::infinity(), horizontalPosition);
    }

    TEST_F(ProfileDataTest, GetProfileSegment_HorizontalPositionLeftOfDikeProfile_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto profileSegment = profileData->GetProfileSegment(0.0);

        // Assert
        ASSERT_EQ(nullptr, profileSegment);
    }

    TEST_F(ProfileDataTest, GetProfileSegment_HorizontalPositionOnFirstDikeProfilePoint_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto profileSegment = profileData->GetProfileSegment(1.0);

        // Assert
        ASSERT_EQ(nullptr, profileSegment);
    }

    TEST_F(ProfileDataTest, GetProfileSegment_HorizontalPositionOnRandomDikeProfilePoint_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto profileSegment = profileData->GetProfileSegment(2.0);

        // Assert
        const auto& startPoint = profileSegment->GetStartPoint();
        const auto& endPoint = profileSegment->GetEndPoint();
        ASSERT_DOUBLE_EQ(1.0, startPoint.GetX());
        ASSERT_DOUBLE_EQ(1.1, startPoint.GetZ());
        ASSERT_DOUBLE_EQ(2.0, endPoint.GetX());
        ASSERT_DOUBLE_EQ(2.2, endPoint.GetZ());
    }

    TEST_F(ProfileDataTest, GetProfileSegment_HorizontalPositionBetweenDikeProfilePoints_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto profileSegment = profileData->GetProfileSegment(1.5);

        // Assert
        const auto& startPoint = profileSegment->GetStartPoint();
        const auto& endPoint = profileSegment->GetEndPoint();
        ASSERT_DOUBLE_EQ(1.0, startPoint.GetX());
        ASSERT_DOUBLE_EQ(1.1, startPoint.GetZ());
        ASSERT_DOUBLE_EQ(2.0, endPoint.GetX());
        ASSERT_DOUBLE_EQ(2.2, endPoint.GetZ());
    }

    TEST_F(ProfileDataTest, GetProfileSegment_HorizontalPositionOnLastDikeProfilePoint_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto profileSegment = profileData->GetProfileSegment(3.0);

        // Assert
        const auto& startPoint = profileSegment->GetStartPoint();
        const auto& endPoint = profileSegment->GetEndPoint();
        ASSERT_DOUBLE_EQ(2.0, startPoint.GetX());
        ASSERT_DOUBLE_EQ(2.2, startPoint.GetZ());
        ASSERT_DOUBLE_EQ(3.0, endPoint.GetX());
        ASSERT_DOUBLE_EQ(3.3, endPoint.GetZ());
    }

    TEST_F(ProfileDataTest, GetProfileSegment_HorizontalPositionRightOfDikeProfile_ExpectedValue)
    {
        // Setup
        const auto profileData = CreateDefaultProfileData();

        // Call
        const auto profileSegment = profileData->GetProfileSegment(4.0);

        // Assert
        ASSERT_EQ(nullptr, profileSegment);
    }
}

File: Test/DiKErnel.Integration.Test/Data/ProfileDataTest.cs
// Copyright (C) Stichting Deltares and State of the Netherlands 2023. All rights reserved.
//
// This file is part of DiKErnel.
//
// This program is free software; you can redistribute it and/or modify it under the terms of
// the GNU Lesser General Public License as published by the Free Software Foundation; either
// version 3 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License along with this
// program; if not, see <https://www.gnu.org/licenses/>.
//
// All names, logos, and references to "Deltares" are registered trademarks of Stichting
// Deltares and remain full property of Stichting Deltares at all times. All rights reserved.

using System.Collections.Generic;
using DiKErnel.Core.Data;
using DiKErnel.Integration.Data;
using DiKErnel.TestUtil;
using NUnit.Framework;

namespace DiKErnel.Integration.Test.Data
{
    [TestFixture]
    internal sealed class ProfileDataTest
    {
        private const double tolerance = 1e-15;

        [Test]
        public void GetVerticalHeight_HorizontalPositionLeftOfDikeProfile_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            double verticalHeight = profileData.GetVerticalHeight(0);

            // Assert
            Assert.That(verticalHeight, Is.NaN);
        }

        [Test]
        public void GetVerticalHeight_HorizontalPositionOnFirstDikeProfilePoint_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            double verticalHeight = profileData.GetVerticalHeight(1);

            // Assert
            Assert.That(verticalHeight, Is.EqualTo(1.1));
        }

        [Test]
        public void GetVerticalHeight_HorizontalPositionOnRandomDikeProfilePoint_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            double verticalHeight = profileData.GetVerticalHeight(2);

            // Assert
            Assert.That(verticalHeight, Is.EqualTo(2.2));
        }

        [Test]
        public void GetVerticalHeight_HorizontalPositionOnLastDikeProfilePoint_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            double verticalHeight = profileData.GetVerticalHeight(3);

            // Assert
            Assert.That(verticalHeight, Is.EqualTo(3.3));
        }

        [Test]
        public void GetVerticalHeight_HorizontalPositionBetweenAscendingDikeProfilePoints_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            double verticalHeight = profileData.GetVerticalHeight(1.5);

            // Assert
            Assert.That(verticalHeight, Is.EqualTo(1.65).Within(tolerance));
        }

        [Test]
        public void GetVerticalHeight_HorizontalPositionBetweenDescendingDikeProfilePoints_ExpectedValue()
        {
            // Setup
            var profilePoint1 = new ProfilePoint(1, 3.3);
            var profilePoint2 = new ProfilePoint(2, 2.2);
            var profilePoint3 = new ProfilePoint(3, 1.1);

            var profileSegmentsWithDescendingHeight = new List<ProfileSegment>
            {
                new ProfileSegment(profilePoint1, profilePoint2, Random.NextDouble()),
                new ProfileSegment(profilePoint2, profilePoint3, Random.NextDouble())
            };

            var profileData = new ProfileData(Random.NextDouble(), profileSegmentsWithDescendingHeight,
                                              new List<CharacteristicPoint>(), CreateDefaultForeshore());

            // Call
            double verticalHeight = profileData.GetVerticalHeight(2.5);

            // Assert
            Assert.That(verticalHeight, Is.EqualTo(1.65).Within(tolerance));
        }

        [Test]
        public void GetVerticalHeight_HorizontalPositionRightOfDikeProfile_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            double verticalHeight = profileData.GetVerticalHeight(4);

            // Assert
            Assert.That(verticalHeight, Is.NaN);
        }

        [Test]
        public void GetHorizontalPosition_VerticalHeightBelowDikeProfile_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            double horizontalPosition = profileData.GetHorizontalPosition(0);

            // Assert
            Assert.That(horizontalPosition, Is.NaN);
        }

        [Test]
        public void GetHorizontalPosition_VerticalHeightOnFirstDikeProfilePoint_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            double horizontalPosition = profileData.GetHorizontalPosition(1.1);

            // Assert
            Assert.That(horizontalPosition, Is.EqualTo(1));
        }

        [Test]
        public void GetHorizontalPosition_VerticalHeightOnRandomDikeProfilePoint_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            double horizontalPosition = profileData.GetHorizontalPosition(2.2);

            // Assert
            Assert.That(horizontalPosition, Is.EqualTo(2));
        }

        [Test]
        public void GetHorizontalPosition_VerticalHeightOnLastDikeProfilePoint_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            double horizontalPosition = profileData.GetHorizontalPosition(3.3);

            // Assert
            Assert.That(horizontalPosition, Is.EqualTo(3));
        }

        [Test]
        public void GetHorizontalPosition_VerticalHeightBetweenProfilePoints_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            double horizontalPosition = profileData.GetHorizontalPosition(1.65);

            // Assert
            Assert.That(horizontalPosition, Is.EqualTo(1.5).Within(tolerance));
        }

        [Test]
        public void GetHorizontalPosition_VerticalHeightAboveDikeProfile_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            double horizontalPosition = profileData.GetHorizontalPosition(4);

            // Assert
            Assert.That(horizontalPosition, Is.NaN);
        }

        [Test]
        public void GetProfileSegment_HorizontalPositionLeftOfDikeProfile_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            ProfileSegment profileSegment = profileData.GetProfileSegment(0);

            // Assert
            Assert.That(profileSegment, Is.Null);
        }

        [Test]
        public void GetProfileSegment_HorizontalPositionOnFirstDikeProfilePoint_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            ProfileSegment profileSegment = profileData.GetProfileSegment(1);

            // Assert
            Assert.That(profileSegment, Is.Null);
        }

        [Test]
        public void GetProfileSegment_HorizontalPositionOnRandomDikeProfilePoint_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            ProfileSegment profileSegment = profileData.GetProfileSegment(2);

            // Assert
            Assert.That(profileSegment.StartPoint.X, Is.EqualTo(1));
            Assert.That(profileSegment.StartPoint.Z, Is.EqualTo(1.1));
            Assert.That(profileSegment.EndPoint.X, Is.EqualTo(2));
            Assert.That(profileSegment.EndPoint.Z, Is.EqualTo(2.2));
        }

        [Test]
        public void GetProfileSegment_HorizontalPositionBetweenDikeProfilePoints_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            ProfileSegment profileSegment = profileData.GetProfileSegment(1.5);

            // Assert
            Assert.That(profileSegment.StartPoint.X, Is.EqualTo(1));
            Assert.That(profileSegment.StartPoint.Z, Is.EqualTo(1.1));
            Assert.That(profileSegment.EndPoint.X, Is.EqualTo(2));
            Assert.That(profileSegment.EndPoint.Z, Is.EqualTo(2.2));
        }

        [Test]
        public void GetProfileSegment_HorizontalPositionOnLastDikeProfilePoint_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            ProfileSegment profileSegment = profileData.GetProfileSegment(3);

            // Assert
            Assert.That(profileSegment.StartPoint.X, Is.EqualTo(2));
            Assert.That(profileSegment.StartPoint.Z, Is.EqualTo(2.2));
            Assert.That(profileSegment.EndPoint.X, Is.EqualTo(3));
            Assert.That(profileSegment.EndPoint.Z, Is.EqualTo(3.3));
        }

        [Test]
        public void GetProfileSegment_HorizontalPositionRightOfDikeProfile_ExpectedValue()
        {
            // Setup
            ProfileData profileData = CreateDefaultProfileData();

            // Call
            ProfileSegment profileSegment = profileData.GetProfileSegment(4);

            // Assert
            Assert.That(profileSegment, Is.Null);
        }

        private static ProfileData CreateDefaultProfileData()
        {
            var profilePoint1 = new ProfilePoint(1, 1.1);
            var profilePoint2 = new ProfilePoint(2, 2.2);
            var profilePoint3 = new ProfilePoint(3, 3.3);

            var profileSegments = new List<ProfileSegment>
            {
                new ProfileSegment(profilePoint1, profilePoint2, Random.NextDouble()),
                new ProfileSegment(profilePoint2, profilePoint3, Random.NextDouble())
            };

            return new ProfileData(Random.NextDouble(), profileSegments, new List<CharacteristicPoint>(), CreateDefaultForeshore());
        }

        private static Foreshore CreateDefaultForeshore()
        {
            return new Foreshore(0.004, -4);
        }
    }
}